src/systems/filecoin_markets/retrieval_market/_index.md

```diff
@@ -1,23 +1,75 @@
+
+The protocol will evolve to piggy back on the Data Transfer system and Graphsync to handle transfer and verification, and to support arbitrary selectors, and to reduce roundtrips. The Data Transfer System will be augmented to support pausing/resuming and sending intermediate vouchers to facilitate this
+
+The underlying protocols will change, but the API interfaces for the client & provider will not change
+
+# Deal Flow (V0)
+
+{{< diagram src="retrieval_flow_v0.mmd.svg" title="Retrieval Flow - V0" >}}
+
+The baseline version of proposing and accepting a deal will work as follows:
+
+- The client finds a provider of a given piece with FindProviders.
+- The client queries a provider to see if it meets its retrieval criteria (via Query Protocol)
+- The client creates a payment channel as neccesary and a lane, ensures there are free funds in the channel
+- The client sends a RetrievalDealProposal to the retrieval miner. (via RetrievalProtocol)
+- The providers validates the proposal and rejects it if it is invalid: invalid price per byte, invalid payment channel are reasons for rejection
+- If the request is valid, the provider responds to with an accept message
+- The provider may also send a small number of blocks over the protocol before requiring payment
+- The client consumes blocks over the retrieval protocol and manually verifies them
+- When the requires payment to proceed, it sends a request for payment and does not send any more blocks
+- The client responds with a payment (or cancels the request)
+- The provider resumes sending blocks
+- The client consumes blocks until payment is required again
+- The process continues till the end of the query
+
+# Deal Flow (V1)
+
+{{< diagram src="retrieval_flow_v1.mmd.svg" title="Retrieval Flow - V1" >}}
+
+The evolved protocol for proposing and accepting a deal will work as follows:
+
+- The client finds a provider of a given piece with FindProviders.
+- The client queries a provider to see if it meets its retrieval criteria (via Query Protocol)
+- The client creates a payment channel as neccesary and a lane, ensures there are free funds in the channel
+- The client schedules a Data Transfer Pull Request passing the RetrievalDealProposal as a voucher.
+- The providers validates the proposal and rejects it if it is invalid: invalid price per byte, invalid payment channel are reasons for rejection
+- If the proposal is valid, the provider responds with an accept message and begins monitoring the data transfer process
+- The provider may begin sending a small amount of bytes before first payment
+- When the provider requires payment, it pauses the data transfer and sends a request for payment as an intermediate voucher
+- The client receives the request for payment responds with a payment, also sent as an intermediate voucher (or cancels the request)
+- The provider unpauses the request and data resumes sending and the process continues till the end of the query
+
+### Bootstrapping Trust
```

Comment:
So this is where I think we can both simplify and add some stuff. Namely:
- I am not sure that enabling a mid-deal change in modalities (specifically price change) is a good idea. This may lead to clients getting held up with half their data downloaded and having to choose between an expensive tear down or racket. Likewise, I wonder if the increment size should not be fixed in terms of the deal (ie front load most of this in the negotiation and then the actual process is just formulaic).
- I would remove choice between pre and post-payment. Let's pick one. Probably favor the client, i.e. I would opt for postpayment given in the first few transfers the data will be unusable as is whereas the payment would not be (for prepayment)
- Finally, I think we are missing some added trust criteria:
    - collateral for either party: the client should potentially lock up negotiated funds on chain and have the vouchers withdraw from there; the provider should lock up a potential cancellation fee in case something goes wrong (all this entails they negotiate ahead of opening a payment channel)
    - timeouts: both for sending data and sending payments. One should not be able to DoS retrieval by simply no longer participating. Both parties should agree to a timeout after which the deal is cancelled and one pays the other a cancellation fee
    - increment size: finally it may be worth having them agree out-of-band on payment frequency and size of data increments (this is my first point again)

Reply:
I agree on just fixing the requirement on post-payment is best. There is an additional benefit to this: because data is sent in IPLD blocks, the size of which are not known ahead (and which can't be split in at least existing protocols), in a prepayment situation you can't make an agreement for exactly n-bytes.

The one possible downside, and I wonder whether you feel this merits a special case, is the end of the retrieval. Here all the motivation (I think) is for the client, once they've received all the data, to short change the miner. I think this "switch" was sort of what I was trying to get at with all the options.

I'm fine with out of band agreements about payment frequency. The only challenge again is the inexact nature of sending IPLD blocks... you're not just sending a byte stream, so it's tricky to divide things up into exact amounts.

Reply:
Good point on the IPLD blocks. Maybe then the frequency is overkill until V1? Or it can be a check for approximate frequency..?

FOr the downside, alas I don't think we can ever really mitigate it: any "fair exchange" solution like this (tit for tat) ends up having someone going on a limb. It's just as well it be the client. This can be derisked by making increments small enough and will have to be taken care of at another layer of the chain (ie through reputation system).